src/loader/loader.c

```diff
@@ -670,26 +672,26 @@ static void inline do_load()
 	char soname[MAX_SO_NAME_LEN];
 	post_parse_analyze_hook_type old_hook;
 
-	strlcpy(soversion, version, MAX_VERSION_LEN);
-
-	/*
-	 * An inval_relcache callback can be called after previous checks of
-	 * loaded had found it to be false. But the inval_relcache callback may
-	 * load the extension setting it to true. Thus it needs to be rechecked
-	 * here again by the outer call after inval_relcache completes. This is
-	 * double-check locking, in effect.
-	 */
-	if (loaded)
-		return;
+	/* If the right version of the library is already loaded, we will just
+	 * skip the actual loading. If the wrong version of the library is loaded,
+	 * we need to kill the session since it will not be able to continue
+	 * operate. */
+	if (*soversion)
+	{
+		if (strcmp(soversion, version) == 0)
+			return;
+		ereport(FATAL,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("extension has been loaded with another version"),
```

Comment:
Should probably say which extension this concerns since it is non-obvious for the user